algosdk/source_map.py

```diff
@@ -0,0 +1,101 @@
+from typing import Dict, Any, List, Tuple
+
+from algosdk.error import SourceMapVersionError
+
+
+class SourceMap:
+    """
+    Decodes a VLQ-encoded source mapping between PC values and TEAL source code lines.
+    Spec available here: https://sourcemaps.info/spec.html
+
+    Args:
+        source_map (dict(str, Any)): source map JSON from algod
+    """
+
+    def __init__(self, source_map: Dict[str, Any]):
+
+        self.version: int = source_map["version"]
+
+        if self.version != 3:
+            raise SourceMapVersionError(self.version)
+
+        self.sources: List[str] = source_map["sources"]
+
+        if "mapping" in source_map:
+            # For backwards compat until go-algorand has the change then we should remove
+            self.mappings: str = source_map["mapping"]
+        else:
+            self.mappings: str = source_map["mappings"]
+
+        pc_list = [
+            _decode_int_value(raw_val) for raw_val in self.mappings.split(";")
+        ]
+
+        self.pc_to_line: Dict[int, int] = {}
+        self.line_to_pc: Dict[int, List[int]] = {}
+
+        last_line = 0
+        for index, line_delta in enumerate(pc_list):
+            if line_delta is not None:  # be careful for '0' checks!
```

Comment:
nit: would be great if we could have a more elaborated comment on this part, discussing that `line_delta` would be `None` if `line_delta == 0`. if I am understanding it correctly

Reply:
I think this is correct? @algochoi?

This comment made me re-evaluate the logic I had to add pc to the line_to_pc map, I think it was a bug to only append to the map IFF the line_delta was not None

Reply:
The current logic here looks correct to me 👍 

In the new encoding scheme, if the value is explicitly zero or `None`, it should equal the `last_line`.

Reply:
I think you can delete the comment there: `# be careful for '0' checks!` as it might cause further confusion?

Reply:
its changed, the above snip is outdated 👍